cirq/value/tensor_product_state.py

```diff
@@ -0,0 +1,345 @@
+# Copyright 2020 The Cirq developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import abc
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Dict
+
+import numpy as np
+
+from cirq import protocols
+from cirq._doc import document
+
+if TYPE_CHECKING:
+    import cirq
+
+
+class _Named1QState(metaclass=abc.ABCMeta):
+    """Abstract class representing a one-qubit state of note."""
+
+    def on(self, qubit) -> 'TensorProductState':
+        """Associates one qubit with this named state.
+
+        The returned object is a TensorProductState of length 1.
+        """
+        return TensorProductState({qubit: self})
+
+    def __call__(self, *args, **kwargs):
+        return self.on(*args, **kwargs)
+
+    @abc.abstractmethod
+    def state_vector(self):
```

Comment:
This is an interesting idea and I think it hints at the overall reality which is: there are a lot of ways you can unambiguously specify a single qubit state! What I had in mind is that `TensorProductState` can be built with any objects that expose a `state_vector` property. Right now, the only such objects are `_Named1QState`s (because this PR is built with #2781 in mind) and there's utility in having an explicitly enumerated list of possibilities. 

Using a circuit to specify the state could work but I'd argue just because you could doesn't mean you should :) I think calling something the plus state and being done with it is easier than saying "it's the state you get by starting in |0> and doing a Hadamard"! 

Like I said, though, there's ample opportunity to expand the scope of states further once we have a use case